packages/async-rewriter/src/async-writer-babel.spec.ts

```diff
@@ -3368,4 +3368,87 @@ switch(TEST) {
       });
     });
   });
+  describe('recursion', () => {
+    describe('non-async', () => {
+      before(() => {
+        spy = sinon.spy(new SymbolTable([{ db: signatures.Database }, {}], signatures));
+        writer = new AsyncWriter(signatures, spy);
+        output = writer.compile(`
+function f(arg) {
+  if (arg === 'basecase') {
+    return 1;
+  }
+  return f();
+}
+  `);
+      });
+      it('compiles correctly', () => {
+        expect(output).to.equal(`function f(arg) {
+  if (arg === 'basecase') {
+    return 1;
+  }
+
+  return f();
+}`);
+      });
+      it('symbol table final state is correct', () => {
+        expect(skipPath(spy.lookup('f'))).to.deep.equal({
+          type: 'function',
+          returnsPromise: false,
+          returnType: signatures.unknown
+        });
+      });
+    });
+    describe('async', () => {
+      before(() => {
+        spy = sinon.spy(new SymbolTable([{ db: signatures.Database }, {}], signatures));
+        writer = new AsyncWriter(signatures, spy);
+        output = writer.compile(`
+function f(arg) {
+  if (arg === 'basecase') {
+    return db.coll.insertOne({});
+  }
+  return f();
+}
+  `);
+      });
+      it('compiles correctly', () => {
+        expect(output).to.equal(`async function f(arg) {
+  if (arg === 'basecase') {
+    return await db.coll.insertOne({});
+  }
+
+  return f();
```

Comment:
Do we expect an await to be added here?

Reply:
I think if is not tail recursive would probably be necessary. Not sure about that case specifically too.